rfcs/rfc-metaprogramming.md

```diff
@@ -0,0 +1,379 @@
+
+Some of these issues can be mitigated by careful use of the metapackage objects. For instance, a program wishing to expose symbols to user code only within a certain package can obtain the meta-package object for that package using a fixed static string, and then know that any symbol-access methods invoked on that object can only operate on symbols within that one package.
+
+## Examples
+
+(several examples inline in the above code. Should possibly find some more bigger use-cases and write them here.)
+
+## Prototype Implementation
+
+Several CPAN modules provide inspiration for abilities and function names:
+
+* [`Package::Stash`](https://metacpan.org/pod/Package::Stash)
+
+* [`Object::Pad::MOP::Class`](https://metacpan.org/pod/Object::Pad::MOP::Class) and related.
+
+## Future Scope
+
+* Currently this RFC makes no consideration for whatever extra abilities may be needed when the `feature-class` branch is introduced. It is expected that additional abilities will be wanted that can operate on classes and members of them (fields, methods, etc...). An insipration can be taken from `Object::Pad::MOP::Class` and related.
+
+* A tangentially-related topic is that of extended API/semantics for declarative attributes. The existing Perl attributes on variables and subroutines are nowhere near as powerful as the ones that will eventually be required by `feature-class` and are already implemented by `Object::Pad`. Whatever extensions are added to core perl to eventually support this will likely have meta-programming accessor methods associated with these object instances too.
+
+* The set of functions provided here is fairly minimal, and no convenient shortcuts are currently considered. As use-cases arise, it may turn out that common patterns, such as reading the value of a given variable in a given package, turn out to be useful. Wrapper functions around these common behaviours could be added on a case-by-case basis.
+
+## Rejected Ideas
+
+* PadWalker's `peek_my`, `peek_our` and `peek_sub`, `set_closed_over`.
+
+These do not seem useful at the current time.
+
+* B::Hooks::EndOfScope
+
+* B::CompilerPhase::Hook
+
+It may be useful to add support for these kinds of abilities somewhere in core perl, but this metaprogramming RFC does not seem to be the place for them.
+
+## Open Issues
+
+* Currently it is unspecified what the "add"-type functions will do if an existing item is already found under the proposed new name. Do they warn before replacing? Is it an error and the user must delete the old one first? Should we provide a "replace"-type function, for which it is an error for the original *not* to exist? Or maybe the semantic should be "add-or-replace" silently.
+
+* Is there actually a use-case for the "enclosed lexicals" functions? I included them because PadWalker seems to suggest they're useful, but I'm not really aware of anything that would benefit from using them.
```

Comment:
For debugging and stack traces, this might prove useful, though I suspect it could be overwhelming.

My [Data::Dumper::Names](https://metacpan.org/pod/Data::Dumper::Names) uses PadWalker to get the names of lexicals and I've always wanted to extend that to also get the names of package variables.

The [PadWalker reverse dependency list](https://metacpan.org/module/PadWalker/requires) might give you some ideas of other uses.

Reply:
I've been convinced by @haarg that actually anything that peeks into closures is probably rather too fragile for a first-cut. I'll take them all out for now. We can always consider adding that back in at a later date if a need arises.